File: elixir/pig-latin/pig_latin.exs
defmodule PigLatin do
  @doc """
  Given a `phrase`, translate it a word at a time to Pig Latin.

  Words beginning with consonants should have the consonant moved to the end of
  the word, followed by "ay".

  Words beginning with vowels (aeiou) should have "ay" added to the end of the
  word.

  Some groups of letters are treated like consonants, including "ch", "qu",
  "squ", "th", "thr", and "sch".

  Some groups are treated like vowels, including "yt" and "xr".
  """
  @spec translate(phrase :: String.t()) :: String.t()
  def translate(phrase) do
    vowelRegex = ~r/(^|\s+)([aeiou]|yt|yd|xr|xb)(\w+)/
    consonantRegex = ~r/(^|\s+)([^aeiou]*qu|[^aeiou]+)(\w+)/

    vowelReplacement = "\\g{1}\\g{2}\\g{3}ay"
    consonantReplacement = "\\g{1}\\g{3}\\g{2}ay"

    if Regex.match?(vowelRegex, phrase) do
      Regex.replace(vowelRegex, phrase, vowelReplacement)
    else
      Regex.replace(consonantRegex, phrase, consonantReplacement)
    end
  end
end

File: ruby/pig-latin/pig_latin.rb
module PigLatin
  def self.translate(sentence)
    if sentence.match?(VOWEL_REGEX)
      sentence.gsub(VOWEL_REGEX, VOWEL_REPLACEMENT)
    else
      sentence.gsub(CONSONANT_REGEX, CONSONANT_REPLACEMENT)
    end
  end

  VOWEL_REGEX = /(?<begin>^|\s+)(?<vowel>a|e|i|o|u|yt|xr)(?<rest>\w+)/.freeze
  VOWEL_REPLACEMENT = '\k<begin>\k<vowel>\k<rest>ay'.freeze

  CONSONANT_REGEX = /(?<begin>^|\s+)(?<consonant>ch|qu|thr|th|rh|sch|yt|\wqu|\w)(?<rest>\w+)/.freeze
  CONSONANT_REPLACEMENT = '\k<begin>\k<rest>\k<consonant>ay'.freeze

  private_constant :VOWEL_REGEX, :CONSONANT_REGEX, :VOWEL_REPLACEMENT, :CONSONANT_REPLACEMENT
end